pkg/process/metadata/workloadmeta/grpc.go

```diff
@@ -5,12 +5,154 @@
+	config    config.ConfigReader
+	extractor *WorkloadMetaExtractor
+	server    *grpc.Server
+	// The address of the server set by start(). Primarily used for testing. May be nil if start() has not been called.
+	addr net.Addr
+
+	wg sync.WaitGroup
+}
+
+// NewGRPCServer creates a new instance of a GRPCServer
+func NewGRPCServer(config config.ConfigReader, extractor *WorkloadMetaExtractor) *GRPCServer {
+	l := &GRPCServer{
+		config:    config,
+		extractor: extractor,
+		server:    grpc.NewServer(),
+	}
+
+	pbgo.RegisterProcessEntityStreamServer(l.server, l)
+	return l
+}
+
+func (l *GRPCServer) consumeProcessDiff(diff *ProcessCacheDiff) ([]*pbgo.ProcessEventSet, []*pbgo.ProcessEventUnset) {
+	setEvents := make([]*pbgo.ProcessEventSet, len(diff.creation))
+	for i, proc := range diff.creation {
+		setEvents[i] = &pbgo.ProcessEventSet{
+			Pid:      proc.pid,
+			Language: &pbgo.Language{Name: string(proc.language.Name)},
+		}
+	}
+
+	unsetEvents := make([]*pbgo.ProcessEventUnset, len(diff.deletion))
+	for i, proc := range diff.deletion {
+		unsetEvents[i] = &pbgo.ProcessEventUnset{Pid: proc.pid}
+	}
+
+	return setEvents, unsetEvents
+}
+
+// Start starts the GRPCServer to listen for new connections
+func (l *GRPCServer) Start() error {
+	log.Info("Starting Process Entity WorkloadMeta gRPC server")
+	listener, err := getListener(l.config)
+	if err != nil {
+		return err
+	}
+	l.addr = listener.Addr()
+	log.Info("Process Entity WorkloadMeta gRPC server has started listening on", listener.Addr().String())
+
+	l.wg.Add(1)
+	go func() {
+		defer l.wg.Done()
+
+		err = l.server.Serve(listener)
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+
+	return nil
 }
 
-var _ mockableGrpcListener = (*noopGRPCListener)(nil)
+// Stop stops and cleans up resources allocated by the GRPCServer
+func (l *GRPCServer) Stop() {
+	log.Info("Stopping Process Entity WorkloadMeta gRPC server")
+	l.server.Stop()
+	l.wg.Wait()
+	log.Info("Process Entity WorkloadMeta gRPC server stopped")
+}
+
+// StreamEntities streams Process Entities collected through the WorkloadMetaExtractor
+func (l *GRPCServer) StreamEntities(_ *pbgo.ProcessStreamEntitiesRequest, out pbgo.ProcessEntityStream_StreamEntitiesServer) error {
+	// When connection is created, send a snapshot of all processes detected on the host so far as "SET" events
+	procs, snapshotVersion := l.extractor.GetAllProcessEntities()
+	setEvents := make([]*pbgo.ProcessEventSet, 0, len(procs))
+	for _, proc := range procs {
+		setEvents = append(setEvents, &pbgo.ProcessEventSet{
+			Pid:      proc.pid,
+			Language: &pbgo.Language{Name: string(proc.language.Name)},
+		})
+	}
+
+	syncMessage := &pbgo.ProcessStreamResponse{
+		EventID:   snapshotVersion,
+		SetEvents: setEvents,
+	}
+	err := out.Send(syncMessage)
+	if err != nil {
+		log.Warnf("error sending process entity event: %s", err)
+		return err
+	}
+
+	// Once connection is established, only diffs (process creations/deletions) are sent to the client
+	for {
```

Comment:
Spoke offline about this, but we should be defensive against the case where there are multiple connections at the same time. I think this could happen if the user is accidentally running multiple instances of the agent.

In this case, what we'd probably want to do is return an error if there is already an open stream.

This can be done in a follow-up though since this feature is not GA in 7.47

Reply:
Completely agree! We're gonna follow-up on this as an improvements for `7.48`.